core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -114,6 +120,20 @@
     private final AtomicInteger throttleRequestCount = new AtomicInteger();
     private final EngineConfig.OpenMode openMode;
     private final AtomicBoolean allowCommits = new AtomicBoolean(true);
+    /**
+     * Refresh listeners. While they are not stored in sorted order they are processed as though they are.
+     */
+    private final LinkedTransferQueue<RefreshListener> refreshListeners = new LinkedTransferQueue<>();
```

Comment:
ok I tell you the truth, I don't think this should be in here. I think the engine should only maintain `private volatile Translog.Location lastRefreshedLocation;` and `refreshListeners` should live in the `IndexShard`. The index shard should offer a method to prune the list which is blocking. For instance it should look like:

```Java
public void notifyListeners(Location location) {
  // here we can notify the listeners.
  // this method should likely block and should be called from the refresh thread as well as from the indexing threads
}
```

I think we should also add a `ReferenceManager.RefreshListener` but that listener should life in the `IndexShard`.